src/planning.py

```diff
@@ -272,29 +295,33 @@ def _run_low_level_search(task: Task, option_model: _OptionModelBase,
                 discovered_failures[cur_idx] = failure
                 # If we're immediately propagating failures, raise it now.
                 if CFG.sesame_propagate_failures == "immediately":
-                    raise _DiscoveredFailureException("Discovered a failure",
-                                                      failure)
+                    raise _DiscoveredFailureException(
+                        "Discovered a failure", failure,
+                        {"longest_refinement": longest_refinement})
             if not discovered_failures[cur_idx]:
                 traj[cur_idx + 1] = next_state
                 cur_idx += 1
                 # Check atoms against expected atoms_sequence constraint.
                 assert len(traj) == len(atoms_sequence)
                 # The expected atoms are ones that we definitely expect to be
-                # true at this point in the plan. They are not *all* the atoms
-                # that could be true.
+                # true at this point in the refinement. They are not *all* the
+                # atoms that could be true.
                 expected_atoms = {
                     atom
                     for atom in atoms_sequence[cur_idx]
                     if atom.predicate.name != _NOT_CAUSES_FAILURE
                 }
-                # This is equivalent to, but faster than, checking whether
-                # expected_atoms is a subset of utils.abstract(traj[cur_idx],
-                # predicates).
+                # This "if all" statement is equivalent to, but faster than,
+                # checking whether expected_atoms is a subset of
+                # utils.abstract(traj[cur_idx], predicates).
                 if all(atom.holds(traj[cur_idx]) for atom in expected_atoms):
                     can_continue_on = True
-                    if cur_idx == len(skeleton):  # success!
-                        result = plan
-                        return result
+                    # Update the longest_refinement found so far.
+                    if cur_idx > len(longest_refinement):
+                        longest_refinement = list(options[:cur_idx])
+                    # If cur_idx is the length of the skeleton, success!
+                    if cur_idx == len(skeleton):
+                        return longest_refinement
```

Comment:
note that at this point (and only this point), `longest_refinement` will be the same as `options`

Reply:
we'll probably have to undo this change if you accept my comment above